pkg/zcapld/zcap.go

```diff
@@ -0,0 +1,208 @@
+	"fmt"
+	"reflect"
+
+	"github.com/hyperledger/aries-framework-go/pkg/doc/verifiable"
+)
+
+const (
+	proofPurposeField         = "proofPurpose"
+	proofCapabilityChainField = "capabilityChain"
+)
+
+// Capability is a ZCAP.
+type Capability struct {
+	ID               string             `json:"id"`
+	Invoker          string             `json:"invoker"`
+	Controller       string             `json:"controller"`
+	Delegator        string             `json:"delegator"`
+	Parent           string             `json:"parentCapability"`
+	Nonce            string             `json:"nonce"`
+	AllowedAction    []string           `json:"allowedAction"`
+	InvocationTarget InvocationTarget   `json:"invocationTarget"`
+	Proof            []verifiable.Proof `json:"proof"`
+}
+
+// InvocationTarget is the target on which the capability applies.
+type InvocationTarget struct {
+	ID   string
+	Type string
+}
+
+// Invokers are this capability's entities authorized to invoke the invocation target.
+func (c *Capability) Invokers() ([]string, error) {
+	// if neither an invoker, controller, nor id is found on the capability then
+	// the capability can not be invoked
+	if c.Invoker == "" && c.Controller == "" && c.ID == "" {
+		return nil, errors.New("invoker not found for capability")
+	}
+
+	// if there's a delegator present and not an invoker, then this capability
+	// was intentionally meant to not be invoked
+	if c.Delegator != "" && c.Invoker == "" {
+		return nil, nil
+	}
+
+	invoker := c.Invoker
+
+	if invoker == "" {
+		invoker = c.Controller
+	}
+
+	if invoker == "" {
+		invoker = c.ID
+	}
+
+	// TODO revisit datatypes of invoker and controller. ocapld.js accounts for any of them to be arrays.
+	return []string{invoker}, nil
+}
+
+// ValidateCapabilityChain validates the capability chain list, ensuring, for instance, it contains only
+// IDs except possibly last entry (which can be a full embedded capability),
+// that all IDs are all absolute URLs, and that it contains no cycles.
+// https://github.com/digitalbazaar/ocapld.js/blob/8a54398162837b1cf52c82978bc8127e52d02974/lib/utils.js#L299
+func (c *Capability) ValidateCapabilityChain() error {
+	chain, err := c.CapabilityChain()
+	if err != nil {
+		return fmt.Errorf("failed to get capabilityChain: %w", err)
+	}
+
+	// TODO no logic for this, no test cases either
+	if len(chain) == 0 {
+		return nil
+	}
+
+	uniqueLinks := make(map[string]*struct{})
+
+	for i := range chain {
+		link := chain[i]
+
+		id, ok := link.(string)
+		if !ok {
+			return errors.New("embedded capabilities in capabilityChain not supported yet")
+		}
+
+		uniqueLinks[id] = nil
+	}
+
+	uniqueLinks[c.ID] = nil
+
+	if len(uniqueLinks) != len(chain)+1 {
+		return errors.New("the capability chain contains a cycle")
+	}
+
+	return nil
+}
+
+// CapabilityChain is the chain of capabilities of this zcap.
+func (c *Capability) CapabilityChain() ([]interface{}, error) {
+	// root capability has no chain
+	if c.Parent == "" {
+		return nil, nil
```

Comment:
return nil nil error here

Reply:
The reference implementation does not consider this an error.

Reply:
https://github.com/digitalbazaar/ocapld.js/blob/8a54398162837b1cf52c82978bc8127e52d02974/lib/utils.js#L196-L199